src/apps/chifra/pkg/cache/read.go

```diff
@@ -0,0 +1,813 @@
+}
+
+func ReadLog(reader *bufio.Reader) (log *types.SimpleLog, err error) {
+	log = &types.SimpleLog{}
+	read := createReadFn(reader)
+	header := &cacheHeader{}
+	err = readCacheHeader(reader, header)
+	if err != nil {
+		return
+	}
+	err = validateHeader(header)
+	if err != nil {
+		return
+	}
+
+	err = readAddress(reader, &log.Address)
+	if err != nil {
+		return
+	}
+
+	err = read(&log.LogIndex)
+	if err != nil {
+		return
+	}
+
+	err = readFromArray(reader, &log.Topics, makeArrayItemRead(reader, readHash))
+	if err != nil {
+		return
+	}
+
+	err = readString(reader, &log.Data)
+	if err != nil {
+		return
+	}
+
+	articulatedLog, err := ReadFunction(reader)
+	if err != nil {
+		return
+	}
+	log.ArticulatedLog = articulatedLog
+
+	return
+}
+
+func ReadFunction(reader *bufio.Reader) (function *types.SimpleFunction, err error) {
+	function = &types.SimpleFunction{}
+	read := createReadFn(reader)
+	header := &cacheHeader{}
+	err = readCacheHeader(reader, header)
+	if err != nil {
+		return
+	}
+	err = validateHeader(header)
+	if err != nil {
+		return
+	}
+
+	err = readString(reader, &function.Name)
+	if err != nil {
+		return
+	}
+
+	err = readString(reader, &function.FunctionType)
+	if err != nil {
+		return
+	}
+
+	err = readString(reader, &function.AbiSource)
+	if err != nil {
+		return
+	}
+
+	err = read(&function.Anonymous)
+	if err != nil {
+		return
+	}
+
+	err = read(&function.Constant)
+	if err != nil {
+		return
+	}
+
+	err = readString(reader, &function.StateMutability)
+	if err != nil {
+		return
+	}
+
+	err = readString(reader, &function.Signature)
+	if err != nil {
+		return
+	}
+
+	err = readString(reader, &function.Encoding)
+	if err != nil {
+		return
+	}
+
+	err = readFromArray(reader, &function.Inputs, ReadParameter)
+	if err != nil {
+		return
+	}
+
+	err = readFromArray(reader, &function.Outputs, ReadParameter)
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+func ReadParameter(reader *bufio.Reader) (param *types.SimpleParameter, err error) {
+	param = &types.SimpleParameter{}
+	read := createReadFn(reader)
+	header := &cacheHeader{}
+	err = readCacheHeader(reader, header)
+	if err != nil {
+		return
+	}
+	err = validateHeader(header)
+	if err != nil {
+		return
+	}
+
+	err = readString(reader, &param.ParameterType)
+	if err != nil {
+		return
+	}
+
+	err = readString(reader, &param.Name)
+	if err != nil {
+		return
+	}
+
+	err = readString(reader, &param.StrDefault)
+	if err != nil {
+		return
+	}
+
+	err = readString(reader, &param.Value)
+	if err != nil {
+		return
+	}
+
+	err = read(&param.Indexed)
+	if err != nil {
+		return
+	}
+
+	err = readString(reader, &param.InternalType)
+	if err != nil {
+		return
+	}
+
+	err = readFromArray(reader, &param.Components, ReadParameter)
+	if err != nil {
+		return
+	}
+
+	err = read(&param.Unused)
+	if err != nil {
+		return
+	}
+
+	err = read(&param.IsFlags)
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+func ReadTrace(reader *bufio.Reader) (trace *types.SimpleTrace, err error) {
```

Comment:
Eventually, we'll move this into the auto-generated type specific file. Please add a TODO, but don't move this yet.